File: typical90/007_CP_Classes.py
N = int(input())
A = sorted(list(map(int, input().split())))


def binary_search(B: int, A: list):
    ub = len(A)
    lb = 0
    while ub - lb > 1:
        mid = int((lb + ub) // 2)
        if B >= A[mid]:
            lb = mid
        else:
            ub = mid
    return lb


def get_distance(B: int, lb: int, A: list):
    if lb == len(A) - 1:
        return abs(B - A[lb])
    elif lb == len(A):
        return abs(B - A[lb - 1])
    else:
        return min(abs(B - A[lb]), abs(A[lb + 1] - B))


Q = int(input())
for i in range(Q):
    B = int(input())
    lb = binary_search(B, A)
    print(get_distance(B, lb, A))

File: typical90/007_CP_Classes.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int binary_search(vector<int> A, int B);
int calc_distance(vector<int> A, int lb, int B);

int main() {
  int N;
  cin >> N;
  vector<int> A(N);
  for (int i = 0; i < N; i++) {
    cin >> A.at(i);
  }
  sort(A.begin(), A.end());
  int Q;
  cin >> Q;
  int B, ans, lb;
  for (int j = 0; j < Q; j++) {
    cin >> B;
    lb = binary_search(A, B);
    ans = calc_distance(A, lb, B);
    cout << ans << endl;
  }
}

int binary_search(vector<int> A, int B) {
  int ub = A.size(), lb = 0, mid;
  for (int i = 0; i < (int)A.size(); i++) {
    if (ub - lb <= 1) {
      break;
    }

    mid = (ub + lb) / 2;
    if (B >= A.at(mid)) {
      lb = mid;
    } else {
      ub = mid;
    }
  }
  return lb;
}

int calc_distance(vector<int> A, int lb, int B) {
  if (lb == (int)A.size()) {
    return abs(B - A.at(lb - 1));
  } else if (lb == (int)A.size() - 1) {
    return abs(B - A.at(lb));
  } else {
    return min(abs(B - A.at(lb)), abs(B - A.at(lb + 1)));
  }
}
